flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,437 @@
+        }
+    }
+
+    // An operator that splits a global batch into evenly-sized local batches, and distributes them
+    // to downstream operator.
+    private static class GlobalBatchSplitter
+            implements FlatMapFunction<DenseVector[], DenseVector[]> {
+        private final int downStreamParallelism;
+
+        private GlobalBatchSplitter(int downStreamParallelism) {
+            this.downStreamParallelism = downStreamParallelism;
+        }
+
+        @Override
+        public void flatMap(DenseVector[] values, Collector<DenseVector[]> collector) {
+            // Calculate the batch sizes to be distributed on each subtask.
+            List<Integer> sizes = new ArrayList<>();
+            for (int i = 0; i < downStreamParallelism; i++) {
+                int start = i * values.length / downStreamParallelism;
+                int end = (i + 1) * values.length / downStreamParallelism;
+                sizes.add(end - start);
+            }
+
+            int offset = 0;
+            for (Integer size : sizes) {
+                collector.collect(Arrays.copyOfRange(values, offset, offset + size));
+                offset += size;
+            }
+        }
+    }
+
+    private static class GlobalBatchCreator
+            implements AllWindowFunction<DenseVector, DenseVector[], GlobalWindow> {
+        @Override
+        public void apply(
+                GlobalWindow timeWindow,
+                Iterable<DenseVector> iterable,
+                Collector<DenseVector[]> collector) {
+            List<DenseVector> points = IteratorUtils.toList(iterable.iterator());
+            collector.collect(points.toArray(new DenseVector[0]));
+        }
+    }
+
+    /**
+     * Sets the initial model data of the online training process with the provided model data
+     * table.
+     *
+     * <p>This would override the effect of previously invoked {@link #setInitialModelData(Table)}
+     * or {@link #setRandomCentroids(StreamTableEnvironment, int, int, double)}.
+     */
+    public OnlineKMeans setInitialModelData(Table initModelDataTable) {
+        this.initModelDataTable = initModelDataTable;
+        return this;
+    }
+
+    /**
+     * Sets the initial model data of the online training process with randomly created centroids.
+     *
+     * <p>This would override the effect of previously invoked {@link #setInitialModelData(Table)}
+     * or {@link #setRandomCentroids(StreamTableEnvironment, int, int, double)}.
+     *
+     * @param tEnv The stream table environment to create the centroids in.
+     * @param dim The dimension of the centroids to create.
+     * @param k The number of centroids to create.
+     * @param weight The weight of the centroids to create.
+     */
+    public OnlineKMeans setRandomCentroids(
+            StreamTableEnvironment tEnv, int dim, int k, double weight) {
+        StreamExecutionEnvironment env = ((StreamTableEnvironmentImpl) tEnv).execEnv();
```

Comment:
While I believe we will have public API on Table to expose its env, I am not sure we will also have public API on `StreamTableEnvironmentImpl` to expose its StreamExecutionEnvironment. 

It will be better to reduce the use of internal APIs if possible. How about we just pass `StreamExecutionEnvironment` as the method's parameter? This would also be more consistent with `load(...)`.

Reply:
OK. I'll make the change.

One small concern is that Flink ML should depend on the Table API, but now it seems that Flink ML also depends on DataStream API.

Reply:
You are right.

Functionality-wise, it seems for Flink ML to depend on both Table and DataStream, because Flink runtime already allows Table to be converted back-and-forth from a DataStream.

I do agree that it seems cleaner to consistently require table environment in our public API. Changing this would require us to file a FLIP. This does not cause real problem, I think it is OK to still use StreamExecutionEnvironment for now.